File: New_Year_Resolution.py
from typing import List


class Solution:
    def isPossible(self, N : int, coins : List[int]) -> bool:
        total = sum(coins)
        possibility = [[False] * (total + 1) for _ in range(N + 1)]
        for i in range(N + 1):
            possibility[i][0] = True
        for i in range(1, total + 1):
            possibility[0][i] = False
        for i in range(1, N + 1):
            for j in range(1, total + 1):
                if j < coins[i - 1]:
                    possibility[i][j] = possibility[i - 1][j]
                if j >= coins[i - 1]:
                    possibility[i][j] = possibility[i - 1][j] or possibility[i - 1][j - coins[i - 1]]
        for i in range(1, total + 1):
            if (i % 20 == 0 or i % 24 == 0 or i == 2024) and possibility[N][i]:
                return True
        return False

File: New_Year_Resolution.java
class Solution{
    boolean isPossible(int N, int[] coins){
    int total = Arrays.stream(coins).sum();
    boolean[][] possibility = new boolean[N + 1][total + 1];
    for (int i = 0; i <= N; i++) {
        possibility[i][0] = true;
    }
    for (int i = 1; i <= total; i++) {
        possibility[0][i] = false;
    }
    for (int idx = 1; idx <= N; idx++) {
        for (int sum = 1; sum <= total; sum++) {
            if (sum < coins[idx - 1]) {
                possibility[idx][sum] = possibility[idx - 1][sum];
            }
            if (sum >= coins[idx - 1]) {
                possibility[idx][sum] = possibility[idx - 1][sum] || possibility[idx - 1][sum - coins[idx - 1]];
            }
        }
    }
    for (int sum = 1; sum <= total; sum++) {
        if ((sum % 20 == 0 || sum % 24 == 0 || sum == 2024) && possibility[N][sum]) {
            return true;
        }
    }
    return false;
}
}
